lib/srv/ctx.go

```diff
@@ -630,6 +630,92 @@ func (c *ServerContext) String() string {
 	return fmt.Sprintf("ServerContext(%v->%v, user=%v, id=%v)", c.Conn.RemoteAddr(), c.Conn.LocalAddr(), c.Conn.User(), c.id)
 }
 
+// ExecCommand takes a *ServerContext and extracts the parts needed to create
+// an *execCommand which can be re-sent to Teleport.
+func (c *ServerContext) ExecCommand() (*execCommand, error) {
+	var pamEnabled bool
+	var pamServiceName string
+
+	// If this code is running on a node, check if PAM is enabled or not.
+	if c.srv.Component() == teleport.ComponentNode {
+		conf, err := c.srv.GetPAM()
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+		pamEnabled = conf.Enabled
+		pamServiceName = conf.ServiceName
+	}
+
+	// If the identity has roles, extract the role names.
+	var roleNames []string
+	if len(c.Identity.RoleSet) > 0 {
+		roleNames = c.Identity.RoleSet.RoleNames()
+	}
+
+	// Create the execCommand that will be send to the child process.
+	return &execCommand{
+		Command:               c.ExecRequest.GetCommand(),
+		Username:              c.Identity.TeleportUser,
+		Login:                 c.Identity.Login,
+		Roles:                 roleNames,
+		Terminal:              c.termAllocated || c.ExecRequest.GetCommand() == "",
+		RequestType:           c.request.Type,
+		PermitUserEnvironment: c.srv.PermitUserEnvironment(),
+		Environment:           buildEnvironment(c),
+		PAM:                   pamEnabled,
+		ServiceName:           pamServiceName,
+		IsTestStub:            c.IsTestStub,
+	}, nil
+}
+
+// buildEnvironment constructs a list of environment variables from
+// cluster information.
+func buildEnvironment(ctx *ServerContext) []string {
+	var env []string
+
+	// Apply environment variables passed in from client.
+	for k, v := range ctx.env {
+		env = append(env, fmt.Sprintf("%s=%s", k, v))
+	}
```

Comment:
Will this also contain anything present in `/etc/environment`? Does that get read in earlier?

Reply:
These are environment variables passed by the client. They are saved in the context then applied here.

https://github.com/gravitational/teleport/blob/master/lib/srv/regular/sshserver.go#L1126-L1127
https://github.com/gravitational/teleport/blob/master/lib/srv/regular/sshserver.go#L1235-L1245

The PAM environment variables are applied during the re-exec, you can see that here:

https://github.com/gravitational/teleport/blob/rjones/pam-cgo/lib/srv/exec.go#L367-L368
https://github.com/gravitational/teleport/blob/rjones/pam-cgo/lib/srv/exec.go#L567-L568

Your comment also raised another interesting question, if a duplicate environment variables exist, which one should win? The one that comes from the client or the one from PAM? I checked what OpenSSH does and the one from PAM wins.

In our case since we apply PAM environment variables last, [they should win](https://golang.org/pkg/os/exec/#Cmd):

```
    // If Env contains duplicate environment keys, only the last
    // value in the slice for each duplicate key is used.
    // As a special case on Windows, SYSTEMROOT is always added if
    // missing and not explicitly set to the empty string.
```